### PR TITLE
 Amesos2: Missing explicit instantiations for MUMPS

### PR DESCRIPTION
**Pull request by finkandreas**
Some explicit instantiations are missing. Relevant cmake config options:
```
-DTrilinos_ENABLE_EXPLICIT_INSTANTIATION=ON
-DTrilinos_ENABLE_OpenMP=ON 
-DTrilinos_ENABLE_Amesos2=ON 
-DAmesos2_ENABLE_MUMPS=ON
-DTpetra_INST_OPENMP=ON 
-DTpetra_INST_SERIAL=ON 
```

### CODE REVIEW

File: packages/amesos2/src/Amesos2_MUMPS.cpp [hunk 1]
```diff
@@ -143,8 +171,22 @@ namespace Amesos2 {
 #ifdef HAVE_TPETRA_INST_COMPLEX_DOUBLE
     AMESOS2_MUMPS_LOCAL_INSTANT(std::complex<double>, int, int, NODETYPE)
```
Comment (mhoemmen) [inline packages/amesos2/src/Amesos2_MUMPS.cpp]: Ditto (see above).

File: packages/amesos2/src/Amesos2_MUMPS.cpp [hunk 2]
```diff
@@ -162,9 +204,22 @@ namespace Amesos2 {
...
```
Comment (mhoemmen) [inline packages/amesos2/src/Amesos2_MUMPS.cpp]: Ditto (see above).

File: packages/amesos2/src/Amesos2_MUMPS.cpp [hunk 3]
```diff
@@ -105,8 +105,22 @@ namespace Amesos2 {
 #ifdef HAVE_TPETRA_INST_COMPLEX_DOUBLE
```
[the remark this answers is not in the document]
Comment (finkandreas) [reply packages/amesos2/src/Amesos2_MUMPS.cpp]: It is already guarded by HAVE_TPETRA_INST_INT_INT...

**Comment by mayrmt** [comment]
This PR wants to merge 48 commits to trilinos/develop, though only the last commit [6be0100](/trilinos/Trilinos/pull/4409/commits/6be010010373388a6a188bb39fac470addeeed9f) brings in  changes related to the issue that should be addressed. Is there a reason for the other 47 commits?

File: packages/amesos2/src/Amesos2_MUMPS.cpp (hunk 1 above)
Comment (finkandreas) [reply packages/amesos2/src/Amesos2_MUMPS.cpp]: Already guarded by `TPETRA_INST_HAVE_INT_INT`

File: packages/amesos2/src/Amesos2_MUMPS.cpp (hunk 2 above)
Comment (finkandreas) [reply packages/amesos2/src/Amesos2_MUMPS.cpp]: Already guarded by `TPETRA_INST_HAVE_INT_INT`

File: packages/amesos2/src/Amesos2_MUMPS.cpp [hunk 4]
```diff
@@ -124,8 +138,22 @@ namespace Amesos2 {
...
     AMESOS2_MUMPS_LOCAL_INSTANT(std::complex<double>, int, int, NODETYPE)
```
[the remark this answers is not in the document]
Comment (finkandreas) [reply packages/amesos2/src/Amesos2_MUMPS.cpp]: Already guarded by `TPETRA_INST_HAVE_INT_INT`

**Comment by ndellingwood** [comment]
@mhoemmen I'm happy with the changes if your requests have been addressed. @finkandreas can the PR be updated to squash the 47 extra commits?

File: packages/amesos2/src/Amesos2_MUMPS.cpp [hunk 5]
```diff
@@ -50,6 +50,7 @@
 
 namespace Amesos2 {
 
+// TODO: How does TPETRA_INST_INT_INT influence an Epetra instantiation???
```
Comment (mhoemmen) [inline packages/amesos2/src/Amesos2_MUMPS.cpp]: It's a known issue that if you enable Epetra in Amesos2, then you must enable `GlobalOrdinal=int` in Tpetra.  See #2548.  This is unfortunate and Trilinos needs to fix it, but I think that's out of scope for this PR.

File: packages/amesos2/src/Amesos2_MUMPS.cpp [hunk 6]
```diff
@@ -70,10 +71,61 @@ namespace Amesos2 {
   AMESOS2_SOLVER_TPETRA_INST(MUMPS,std::complex<double>,int,int);
```
Comment (mhoemmen) [inline packages/amesos2/src/Amesos2_MUMPS.cpp]: Not sure if putting colons or angle brackets in macro arguments works quite right.  (Spaces, like in `long long`, definitely do _not_ work right.)  Lifting the `TPETRA_ETI_MANGLING_TYPEDEFS()` macro invocation above this code would let you use the typedefs that the macro generates, like `std_complex0double0` and `longlong`.

There's an easier way for you to do this: use Tpetra's instantiation macros.  Look for `packages/tpetra/core/src/TpetraCore_ETIHelperMacros.h` in your _build_ (not source) directory.  Include it in this file, and try this:
```c++
#define AMESOS2_SOLVER_TPETRA_INST_MUMPS( SC, LO, GO ) \
  AMESOS2_SOLVER_TPETRA_INST(MUMPS, SC, LO, GO)

TPETRA_ETI_MANGLING_TYPEDEFS()

TPETRA_INSTANTIATE_SLG(AMESOS2_SOLVER_TPETRA_INST_MUMPS)
```
That would get rid of all of your uses of `AMESOS2_SOLVER_TPETRA_INST`.  You would still need to use `AMESOS2_SOLVER_EPETRA_INST` explicitly, like you do above.

Comment (finkandreas) [reply packages/amesos2/src/Amesos2_MUMPS.cpp]: Thanks for pointing out the ETI helper macros. I wasn't aware of their existence. This reduced the code to just a few lines in the end. 
I think this macros should be used in all other explicit instantiations too, so new developers will see them easily, when they copy and paste code ;)

**Comment by finkandreas** [comment]
I'm not entirely sure what happened, but now it seems that too many files would be changed... Should I open a new PR for this, to have it clean?

**Comment by srajama1** [comment]
@finkandreas : Looks like a merging issue. If you don't mind, yes please create a new PR.

**Comment by finkandreas** [comment]
Closing this pull request in favor of [this one](https://github.com/trilinos/Trilinos/pull/4455)

**Pull request closed without merging**